Feature/Pool Contract

ecmendenhall opened the pull request:
Opening a PR so we can move the conversation here.

anthonymartin commented on packages/contracts/contracts/mocks/MockCurveDepositZap.sol:
```diff
@@ -0,0 +1,63 @@
+    uint256 transferOut = amount - slippage;
+
+    dai.approve(address(this), transferOut);
+    dai.mint(address(this), transferOut);
```
is there a reason to mint from here?

ecmendenhall replied:
Yeah, this is necessary when we're simulating a withdrawal after earning some yield in the underlying vault. In that scenario, we've earned more LP tokens, so we'll need to withdraw more from this mock than we originally sent it. Minting here makes an implicit assumption that Curve has unlimited liquidity for withdrawals.

ecmendenhall commented on packages/contracts/contracts/Pool.sol:
```diff
@@ -0,0 +1,339 @@
+  function _takePerformanceFee() internal {
+    int256 gain = int256(pricePerPoolToken() - poolTokenHWM);
+    if (gain > 0) {
+      uint256 changeInPricePerToken = uint256(gain);
```
We now know this is positive. We should probably subtract again rather than cast from `int` back to `uint` here.

ecmendenhall replied:
...or just do a `pricePerPoolToken() > poolTokenHWM` comparison and avoid casting anything at all.

ecmendenhall commented on packages/contracts/contracts/Pool.sol:
```diff
@@ -0,0 +1,339 @@
+    uint256 crvLPTokenAmount = _sendToCurve(amount);
+    _sendToYearn(crvLPTokenAmount);
+
+    _reportPoolTokenHWM();
```
Is it possible this can always happen right before `_takeFees`?

anthonymartin commented:
I had some issue testing this contract specifically with the safeApprove methods. On sequential deposits I got this error: `revert SafeERC20: approve from non-zero to non-zero allowance`

see https://github.com/OpenZeppelin/openzeppelin-contracts/issues/2219#issuecomment-622163352

perhaps we should use safeIncreaseAllowance

ecmendenhall commented:
Updates since last review on this:
- Switched to `safeIncreaseAllowance`
- Removed 1:1 share minting
- Added additional Curve interfaces to get virtual price
- Use virtual price for share valuation
- Don't include withdrawal fees in share valuations

ecmendenhall commented on packages/contracts/contracts/Pool.sol:
```diff
@@ -0,0 +1,335 @@
+  address public governance;
+
+  uint256 constant BPS_DENOMINATOR = 10_000;
+  uint256 constant YEARN_PRECISION = 10e17;
```
Both, and neither: this was a bad name when it was called `YEARN_PRECISION`. We removed this constant and inlined these, because ultimately this didn't have anything to do with Yearn—it's the degree of precision we're using in our own division operations.
